### PR TITLE
[spinel] add support for logging crash dumps

### LINKED ISSUE #10048: Request for a new hook that occurs after Host gets RCP capabilities
**Is your feature request related to a problem? Please describe.** A clear and concise description of what the problem is.

RCPs don't have a hook that's executed after the Host finishes initializing RCP capabilities.

In my use case, I'm trying to print some crash logs from the RCP using `LOG_STREAM` spinel messages to the Host after recovering from a crash.

**Describe the solution you'd like** A clear and concise description of what you want to happen.

I think there are many ways to implement something that would address the feature request. Here are a couple independent solutions I have in mind:

- Add a new [Notifier event](https://github.com/openthread/openthread/blob/fa6477197c61a7e69e47e9e8dd9a032cd38a709b/src/core/common/notifier.hpp#L67-L101) `kRcpCapabilitiesSentToHost`.
  - On the RCP, this event would be emitted at the end of [`NcpBase::HandlePropertyGet<SPINEL_PROP_CAPS>()`](https://github.com/openthread/openthread/blob/fa6477197c61a7e69e47e9e8dd9a032cd38a709b/src/ncp/ncp_base.cpp#L2068-L2069) if no error has occurred.
  - This would allow a vendor to use the event to send crash logs or any other logs upon a Host establishing a connection to the RCP using [`ExtensionBase::HandleNotifierEvents()`](https://github.com/openthread/openthread/blob/fa6477197c61a7e69e47e9e8dd9a032cd38a709b/src/core/common/notifier.cpp#L175)
  - Nothing would need to be added to the Host code.

- Add a new spinel capability and property `SPINEL_PROP/CAP_RCP_INIT_COMPLETE`.
  - After the Host gets the RCP capabilities, if a RCP has this capability, the Host sends `SPINEL_PROP_RCP_INIT_COMPLETE` to RCP.
  - The handler for this in `ncp_base_radio.cpp` would call a new `otPlatRcpInitComplete()` platform API.
    - I recognize that @abtink [expressed some concern in my PR](https://github.com/openthread/openthread/pull/9829#discussion_r1478940471) about how `otPlat...()` APIs are intended for OT core modules use. 
    - It looks like [`otPlatMultipanGet/SetActiveInstance()`](https://github.com/search?q=repo%3Aopenthread%2Fopenthread%20otPlatMultipanGetActiveInstance&type=code) is being used in a similar way to how I invision `otPlatRcpInitComplete()` being used. `otPlatMultipanGet/SetActiveInstance()` are only being called in [`NcpBase::HandlePropertyGet<SPINEL_PROP_MULTIPAN_ACTIVE_INTERFACE>()`](https://github.com/openthread/openthread/blob/4737231d64d262eb850e0da729d4b2df130033e1/src/ncp/ncp_base_radio.cpp#L589)
    - Alternatively, a callback could be added to the `NcpBase` class instead of adding a new `otPlat...()` API

**Describe alternatives you've considered** A clear and concise description of any alternative solutions or features you've considered.

- Adding a new spinel capability and property specifically for logging crash dumps
  - I proposed this in https://github.com/openthread/openthread/pull/9829
  - This approach would have allowed for the Host to request the latest crash dump from the RCP
- @zhanglongxia suggested in (https://github.com/openthread/openthread/pull/9829#issuecomment-1925563823) that using [`ExtensionBase::SignalNcpInit()`](https://github.com/openthread/openthread/blob/main/src/core/instance/extension_example.cpp#L79) might be a better approach.
  - I implemented this and upon initial testing using a `diag` command to trigger a fault, this solution works since the RCP crashes and resets while the Host has an existing/established connection to the RCP. See 
    - See log [ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt](https://github.com/openthread/openthread/files/15045096/ot-cli-syslog_ot-rcp-using-vendor-hook-to-print-logs.txt)
 
  - After more thorough testing, I discovered that there are still some scenarios where the RCP can send `LOG_STREAM` spinel messages to the Host before the Host is aware of the RCP capabilities, causing the logs to be missed on the Host.
  - Specifically, this occurs when I configure the RCP to crash every 2 seconds. Since the [`ExtensionBase::SignalNcpInit()`](https://github.com/openthread/openthread/blob/main/src/core/instance/extension_example.cpp#L79) hook is called at the end of the [NcpBase::NcpBase() constructor](https://github.com/openthread/openthread/blob/fa6477197c61a7e69e47e9e8dd9a032cd38a709b/src/ncp/ncp_base.cpp#L361), it's possible for the hook to be called before `ot-cli` has been started on the Host.

**Additional context** Add any other context or screenshots about the feature request here.

### PR DESCRIPTION
**Pull request by lmnotran**
Supersedes #9829
Closes #10048
---

This PR adds a new feature that allows platforms to log crash logs.

### Additions
- `void otPlatLogCrashDump(void)` - API that logs a crash dump using OpenThread Logging APIs
- `SPINEL_PROP_RCP_LOG_CRASH_DUMP` - spinel prop that calls `otPlatLogCrashDump()` when `Set`
- `SPINEL_CAP_RCP_LOG_CRASH_DUMP` - spinel capability denoting that a RCP supports logging crash dumps

### Usage
- For `ot-cli-ftd|mtd`, `otPlatLogCrashDump()` is called at the end of app initialization, before the main loop. See [main.c]
- For Host/RCP setups, during initialization, the Host gets the RCP capabilities. If the RCP has the `SPINEL_CAP_RCP_LOG_CRASH_DUMP` capability, the Host will `Set` the `SPINEL_PROP_RCP_LOG_CRASH_DUMP` property, triggering the RCP to call `otPlatLogCrashDump()`.
  - If RCP Recovery is enabled, the this will also happen once the RCP has been restored

### Example logs

[ot-cli-syslog_ot-rcp-fault-triggered-by-diag-command.md](https://github.com/openthread/openthread/files/15105488/ot-cli-syslog_ot-rcp-fault-triggered-by-diag-command.md)
[ot-cli-syslog_ot-rcp-autofaulting-every-2-seconds.md](https://github.com/openthread/openthread/files/15105489/ot-cli-syslog_ot-rcp-autofaulting-every-2-seconds.md)

### CODE REVIEW

**Comment by lmnotran** [comment]
@zhanglongxia @abtink Could you please review this when you get a chance? Thanks!

File: etc/cmake/options.cmake [hunk 1]
```diff
@@ -233,6 +233,7 @@ ot_option(OT_OTNS OPENTHREAD_CONFIG_OTNS_ENABLE "OTNS")
 ot_option(OT_PING_SENDER OPENTHREAD_CONFIG_PING_SENDER_ENABLE "ping sender" ${OT_APP_CLI})
 ot_option(OT_PLATFORM_BOOTLOADER_MODE OPENTHREAD_CONFIG_PLATFORM_BOOTLOADER_MODE_ENABLE "platform bootloader mode")
 ot_option(OT_PLATFORM_KEY_REF OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE "platform key reference secure storage")
+ot_option(OT_PLATFORM_LOG_CRASH_DUMP OPENTHREAD_CONFIG_PLATFORM_LOG_CRASH_DUMP_ENABLE "platform log crash dump")
```
[the remark this answers is not in the document]
Comment (abtink) [reply etc/cmake/options.cmake:236]: May be useful to enable this in some build config used by Github Action tests/build so to make sure it is validated by CI build checks?

**Review by abtink — COMMENTED** [review]
Thanks @lmnotran. Some suggestions below. Thank you.

File: src/ncp/ncp_base_dispatcher.cpp [hunk 2]
```diff
@@ -479,6 +479,13 @@ NcpBase::PropertyHandler NcpBase::FindSetPropertyHandler(spinel_prop_key_t aKey)
 #if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
         OT_NCP_SET_HANDLER_ENTRY(SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE),
 #endif
+#endif // OPENTHREAD_MTD || OPENTHREAD_FTD
+
+#if OPENTHREAD_CONFIG_PLATFORM_LOG_CRASH_DUMP_ENABLE
+        OT_NCP_SET_HANDLER_ENTRY(SPINEL_PROP_RCP_LOG_CRASH_DUMP),
```
[the remark this answers is not in the document]
Comment (lmnotran) [reply src/ncp/ncp_base_dispatcher.cpp:485]: I believe it is sorted.

```c
SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE = 160
SPINEL_PROP_RCP_LOG_CRASH_DUMP = 178
SPINEL_PROP_CNTR_RESET = 1280
```

**Review by abtink — APPROVED** [review]
Looks great. Thanks. 👍 

One small suggestion below.

File: src/lib/spinel/radio_spinel.cpp [hunk 3]
```diff
@@ -147,6 +149,12 @@ void RadioSpinel::Init(SpinelInterface    &aSpinelInterface,
     SuccessOrExit(error = Get(SPINEL_PROP_HWADDR, SPINEL_DATATYPE_EUI64_S, sIeeeEui64.m8));
     InitializeCaps(supportsRcpApiVersion, supportsRcpMinHostApiVersion);
 
+    if (sSupportsLogCrashDump)
+    {
+        otLogDebgPlat("RCP supports crash dump logging. Requesting crash dump.");
```
[the remark this answers is not in the document]
Comment (lmnotran) [reply src/lib/spinel/radio_spinel.cpp]: 💯 ! Thanks for the suggestion!

**Comment by lmnotran** [comment]
> Looks great. Thanks. 👍 
> 
> One small suggestion below.

Thanks for reviewing, @abtink !

**Review by zhanglongxia — APPROVED** [review]
LGTM

**Comment by lmnotran** [comment]
@jwhui Could you review this soon please? This is blocking a release issue for us. Thanks!

**Review by jwhui — APPROVED** [review]

**Pull request merged**